### PR TITLE
[FIX] fields: writing on a one2many with a domain should not remove a…

### LINKED ISSUE #18438: Adding records to a one2many deletes records if the domain uses a lambda
Impacted versions: 8, 9, not tested in others

Steps to reproduce:

Install [`l10n_es_aeat_mod303`](https://www.odoo.com/apps/modules/9.0/l10n_es_aeat_mod303/) and [`l10n_es_aeat_mod111`](https://www.odoo.com/apps/modules/9.0/l10n_es_aeat_mod111/).

Generate one 303 report and one 111 report. You should ensure both have the same ID to reproduce the bug.

Calculate taxes for one and then for the other.

Current behavior: When the 2nd model generates its taxes, it deletes 1st's.

The offending line is [this one](https://github.com/OCA/l10n-spain/blob/0dfe806dcfddf595d1ab36ee1cba32d19156eb11/l10n_es_aeat/models/l10n_es_aeat_report_tax_mapping.py#L38). See the code:

```python
class L10nEsAeatReportTaxMapping(models.AbstractModel):
    _name = "l10n.es.aeat.report.tax.mapping"
    _inherit = "l10n.es.aeat.report"
    _description = ("Inheritable abstract model to add taxes by code mapping "
                    "in any AEAT report")

    tax_line_ids = fields.One2many(
        comodel_name='l10n.es.aeat.tax.line', inverse_name='res_id',
        domain=lambda self: [("model", "=", self._name)], auto_join=True,
        readonly=True, oldname='tax_lines', string="Tax lines")

    @api.multi
    def calculate(self):
        res = super(L10nEsAeatReportTaxMapping, self).calculate()
        for report in self:
            report.tax_line_ids.unlink()
            # Buscar configuración de mapeo de impuestos
            tax_code_map = self.env['l10n.es.aeat.map.tax'].search(
                [('model', '=', report.number),
                 '|',
                 ('date_from', '<=', report.date_start),
                 ('date_from', '=', False),
                 '|',
                 ('date_to', '>=', report.date_end),
                 ('date_to', '=', False)], limit=1)
            if tax_code_map:
                tax_lines = []
                for map_line in tax_code_map.map_line_ids:
                    tax_lines.append(report._prepare_tax_line_vals(map_line))
                report.tax_line_ids = [(0, 0, x) for x in tax_lines]
        return res
```

Notice that `tax_line_ids` is declared with `domain=lambda self: [("model", "=", self._name)]`, a common pattern when you are doing `one2many` to a flexible model.

Then, later, when doing `report.tax_line_ids = [(0, 0, x) for x in tax_lines]`, it produces an `unlink` of all `l10n.es.aeat.tax.line` where their id match to `report.id` but their `model` does not match the `report._name`.

We tried replacing the offending line with:

```python
for line in tax_lines:
    report.tax_line_ids |= self.env['l10n.es.aeat.tax.line'].new(line)
```

... but it happens the same.

Expected behavior: Adding records should not delete others.

@Tecnativa @rco-odoo

### PR DESCRIPTION
**Pull request by nim-odoo**
…ll lines

Specifically, the command `(6, 0, ids)` should only unlink/detach the lines
that satisfy to the field's domain.

Test is the courtesy of @Yajo

opw-756983
Closes #18438
Closes #18440

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by nim-odoo** [comment]
@Yajo done ;)

File: openerp/osv/fields.py [hunk 1]
```diff
@@ -861,8 +861,13 @@ def set(self, cr, obj, id, field, values, user=None, context=None):
                     # Must use write() to recompute parent_store structure if needed
                     obj.write(cr, user, act[2], {self._fields_id:id}, context=context or {})
                     ids2 = act[2] or [0]
-                    cr.execute('select id from '+_table+' where '+self._fields_id+'=%s and id <> ALL (%s)', (id,ids2))
-                    ids3 = map(lambda x:x[0], cr.fetchall())
+                    # if the o2m has a static domain we must respect it when unlinking
+                    domain = (self._domain(original_obj)
+                              if callable(self._domain) else self._domain)
+                    extra_domain = domain or []
+                    ids3 = obj.search(cr, user, [(self._fields_id,'=',id), ('id','not in',ids2)] + extra_domain, context=context)
```
Comment (pedrobaeza) [inline openerp/osv/fields.py]: Is this going to affect performance in case no domain is defined?

Comment (nim-odoo) [reply openerp/osv/fields.py]: I'm not sure I understand your concern. The domain will be at least `[(self._fields_id,'=',id), ('id','not in',ids2)]`

Comment (pedrobaeza) [reply openerp/osv/fields.py]: Previously, the query was done directly in SQL, and now is done in ORM. The question is if this affects something in the performance. If it does, maybe we should keep the query in SQL when no domain is found, and only switch to ORM when there's a domain.

Comment (nim-odoo) [reply openerp/osv/fields.py]: Yep, this was a concern. We will test this internally for a couple of days/weeks before merging anything anyway.

Comment (pedrobaeza) [reply openerp/osv/fields.py]: OK, thanks.

**Comment by pedrobaeza** [comment]
Thanks!

**Pull request merged**